### PR TITLE
Z3 4.8.14

### PR DESCRIPTION
**Pull request by mario-bucev**
To get `getAbsFuncDecl` working, we savagely apply a patch on the Z3 sources to export `mkAbs`.
The other changes are mostly whitespace trimming (done by the editor...) and putting `()` to methods expecting them to address the warnings when cross-compiling to Scala 2.13

### CODE REVIEW

**Comment by samarion** [comment]
> To get getAbsFuncDecl working, we savagely apply a patch on the Z3 sources to export mkAbs.

So much violence! :)

If the tests pass, this LGTM. It would probably be worth updating the unmanaged jars in [Inox](https://github.com/epfl-lara/inox/tree/master/unmanaged) and [Stainless](https://github.com/epfl-lara/stainless/tree/master/unmanaged) as well.

**Comment by mario-bucev** [comment]
I had a quick glance at updating the unmanaged jars in Inox and some tests fail, but it should not be too hard to update them.

**Comment by samarion** [comment]
Hmm, failing tests don't sound too good. Is 4.8.12 a stable Z3 version? Which version are we shipping in the z3 binary?

**Comment by mario-bucev** [comment]
It seems so; what's weird is that all tests (in the master branch) that communicate with Z3 through SMT-LIB use the 4.8.12 version and all pass.

**Comment by samarion** [comment]
Which Inox tests are failing with this version? Maybe there's been a subtle API change in the native version which we have to take into account?

Maybe you can try a different Z3 version otherwise?

**Comment by mario-bucev** [comment]
The Z3 Set Extraction test (in `SemanticsSuite`, for `nativez3` and `unrollz3` only) as well as the Model Extraction test (in `SolversSuite`, only for `nativez3`).
I'll investigate the issue and try your suggestion as well.

**Comment by samarion** [comment]
Ah, then it's fairly likely that we're running into bugs in Inox if it's only about model extraction.

**Comment by mario-bucev** [comment]
Z3 is returning a `lambda` which the `Z3Native` does not currently support. Maybe we can do something similar to what is done in `Z3Target`:
https://github.com/epfl-lara/inox/blob/efd76f30e486a89fc4fc33c073a69f3bf2bfd052/src/main/scala/inox/solvers/smtlib/Z3Target.scala#L187-L209

**Comment by samarion** [comment]
Adapting `Z3Native` to handle lambdas like `Z3Target` sounds like a plan.

**Comment by mario-bucev** [comment]
Upgrading to 4.8.14 resolves the problem with the set extraction, as Z3 no longer returns lambdas (seems to be related to the fix of https://github.com/Z3Prover/z3/issues/5604). There are some integration tests that fail though (I did not run 4.8.12 on the integration tests, but they are likely to fail as well).

File: src/main/scala/z3/scala/Z3Context.scala [hunk 1]
```diff
@@ -1024,22 +1029,8 @@ sealed class Z3Context(val config: Map[String, String]) {
     res
   }
 
-  // XXX: this is a HUGE hack to get around the fact that the Z3 api doesn't
-  //      provide a handle to the absolute value function.
   def getAbsFuncDecl(): Z3FuncDecl = {
-    val ast = parseSMTLIB2String("""
```
Comment (gsps) [inline src/main/scala/z3/scala/Z3Context.scala:1030]: For what it's worth, given the `mkAbs` helper you added below it should be possible to make this marginally less hacky by avoiding the parsing. I had to do something similar to get a handle to the built-in if-then-else function here: https://github.com/epfl-lara/inox/blob/2f402f9bf9541cc061558e525fdd53efa5f86a87/src/main/scala/inox/solvers/z3/Z3Native.scala#L478

**Comment by mario-bucev** [comment]
With some small changes, all Inox tests pass. Stainless does not need any adaptations (except for its dependencies of course), and one Bolts benchmark, `gcd`, needs one tiny change.

**Comment by samarion** [comment]
That sounds reasonable. Can you please open a PR with the Inox changes too?

**Comment by mario-bucev** [comment]
Sure! Here it is: https://github.com/epfl-lara/inox/pull/175

**Pull request merged**